### PR TITLE
Task/bws110 delete bucket

### PR DESCRIPTION
**Pull request by Donymak**
https://user-images.githubusercontent.com/86227025/155219421-f18d99e4-a59a-4833-a53a-e6737ddecc32.mp4

### CODE REVIEW

File: frontend/src/components/bs/components/spaces-table/helpers/delete-row-cell/delete-row-cell.tsx [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import styles from './styles.module.scss';
+import deleteIcon from 'assets/img/delete-icon.svg';
+import { BSSpaceDeleteParamsDto } from 'common/types/types';
+
+const DeleteRowCell = (
+  id: string,
+  handleSpaceDelete: (payload: BSSpaceDeleteParamsDto) => void,
```
Comment (what1s1ove) [inline frontend/src/components/bs/components/spaces-table/helpers/delete-row-cell/delete-row-cell.tsx]:
```suggestion
  onSpaceDelete: (id string) => void,
```

File: frontend/src/components/bs/components/spaces-table/helpers/delete-row-cell/delete-row-cell.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+): JSX.Element => {
+  return (
+    <img
```
Comment (what1s1ove) [inline frontend/src/components/bs/components/spaces-table/helpers/delete-row-cell/delete-row-cell.tsx]: should be button with button image

File: frontend/src/components/bs/components/spaces-table/helpers/delete-row-cell/delete-row-cell.tsx [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import styles from './styles.module.scss';
```
Comment (what1s1ove) [inline frontend/src/components/bs/components/spaces-table/helpers/delete-row-cell/delete-row-cell.tsx]: it should be in the `cells` folder

File: frontend/src/components/bs/components/spaces-table/helpers/get-rows.helper.ts [hunk 4]
```diff
@@ -1,18 +1,33 @@
...
-    const { name, createdAt } = item;
+    const { name, createdAt, id } = item;
 
     return {
       [SpacesTableAccessor.SPACE_NAME]: name,
-      [SpacesTableAccessor.CREATION_TIME]: createdAt,
+      [SpacesTableAccessor.CREATION_TIME]: getDistanceToDateNow(
+        new Date(createdAt),
+      ),
+      [SpacesTableAccessor.ACTIONS]: DeleteRowCell(id, handleSpaceDelete),
```
Comment (what1s1ove) [inline frontend/src/components/bs/components/spaces-table/helpers/get-rows.helper.ts]:
```suggestion
      [SpacesTableAccessor.ACTIONS]: DeleteRowCell(id, onSpaceDelete),
```

File: frontend/src/components/bs/components/spaces-table/spaces-table.tsx [hunk 5]
```diff
@@ -1,14 +1,20 @@
 import { FC } from 'react';
 import { useAppSelector, useMemo } from 'hooks/hooks';
+import { BSSpaceDeleteParamsDto } from 'common/types/types';
 import { Table } from 'components/common/common';
 import { getRows, getColumns } from './helpers/helpers';
 
-const SpacesTable: FC = ({ children }) => {
+type Props = {
+  children: JSX.Element;
+  handleSpaceDelete: ({ id }: BSSpaceDeleteParamsDto) => void;
```
Comment (what1s1ove) [inline frontend/src/components/bs/components/spaces-table/spaces-table.tsx]: A8

File: frontend/src/services/bs-api/bs-api.service.ts [hunk 6]
```diff
@@ -39,6 +40,17 @@ class BSApi {
     );
   }
 
+  public deleteSpace(
+    params: BSSpaceDeleteParamsDto,
```
Comment (what1s1ove) [inline frontend/src/services/bs-api/bs-api.service.ts]:
```suggestion
    id: string,
```

File: frontend/src/store/bs/actions.ts [hunk 7]
```diff
@@ -1,10 +1,12 @@
 import { createAsyncThunk } from '@reduxjs/toolkit';
 import {
+  BSSpaceDeleteParamsDto,
   BSSpaceGetRequestParamsDto,
   BSSpaceGetResponseDto,
   AsyncThunkConfig,
 } from 'common/types/types';
 import { ActionType } from './common';
+import { notification } from 'services/services';
```
Comment (what1s1ove) [inline frontend/src/store/bs/actions.ts]:
```suggestion
```
use `extra` in an action

File: frontend/src/store/bs/actions.ts [hunk 8]
```diff
@@ -16,4 +18,18 @@ const loadSpaces = createAsyncThunk<
   return bsApi.loadSpaces(filter);
 });
 
-export { loadSpaces };
+const deleteSpace = createAsyncThunk<
+  BSSpaceDeleteParamsDto,
+  BSSpaceDeleteParamsDto,
+  AsyncThunkConfig
+>(ActionType.DELETE_SPACE, async (payload, { extra }) => {
```
Comment (what1s1ove) [inline frontend/src/store/bs/actions.ts]:
```suggestion
>(ActionType.DELETE_SPACE, async (id, { extra }) => {
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/components/bs/components/spaces-table/helpers/cells/delete-row-cell/delete-row-cell.tsx [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+const DeleteRowCell = (
+  id: string,
+  onSpaceDelete: (id: string) => void,
+): JSX.Element => {
+  return (
+    <button
+      className={styles.button}
+      onClick={(): void => {
+        onSpaceDelete(id);
+      }}
```
Comment (what1s1ove) [inline frontend/src/components/bs/components/spaces-table/helpers/cells/delete-row-cell/delete-row-cell.tsx]:
```suggestion
  const handleSpaceDelete = () => {
    onSpaceDelete(id);
  }
      onClick={handleSpaceDelete}
```
avoid anonymous functions

File: frontend/src/components/bs/components/spaces-table/helpers/cells/delete-row-cell/delete-row-cell.tsx [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    >
+      <img src={deleteIcon} alt={'Delete'} />
```
Comment (what1s1ove) [inline frontend/src/components/bs/components/spaces-table/helpers/cells/delete-row-cell/delete-row-cell.tsx]:
```suggestion
      <img src={deleteIcon} alt="Delete" />
```

File: frontend/src/components/bs/components/spaces-table/spaces-table.tsx [hunk 11]
```diff
@@ -3,12 +3,17 @@ import { useAppSelector, useMemo } from 'hooks/hooks';
 import { Table } from 'components/common/common';
 import { getRows, getColumns } from './helpers/helpers';
 
-const SpacesTable: FC = ({ children }) => {
+type Props = {
+  children: JSX.Element;
```
Comment (what1s1ove) [inline frontend/src/components/bs/components/spaces-table/spaces-table.tsx]:
```suggestion
```
`FC` has this type

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/services/bs-api/bs-api.service.ts [hunk 12]
```diff
@@ -39,6 +39,15 @@ class BSApi {
     );
   }
 
+  public deleteSpace(id: string): Promise<boolean> {
+    return this.#http.load(
+      joinItems(this.#apiPrefix, ApiPath.BS, BSApiPath.SPACES, '/', id),
```
Comment (what1s1ove) [inline frontend/src/services/bs-api/bs-api.service.ts]:
```suggestion
      joinItems(this.#apiPrefix, ApiPath.BS, BSApiPath.SPACES, id),
```

Comment (what1s1ove) [reply frontend/src/services/bs-api/bs-api.service.ts:44]: or add enum

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/services/bs-api/bs-api.service.ts (hunk 12 above)
Comment (Donymak) [reply frontend/src/services/bs-api/bs-api.service.ts]: image
`joinItems(this.#apiPrefix, ApiPath.BS, BSApiPath.SPACES, BSApiPath.ROOT, id)` ?

Comment (what1s1ove) [reply frontend/src/services/bs-api/bs-api.service.ts]: then, add enum for it

File: shared/src/common/enums/api/bs-api-path.enum.ts [hunk 13]
```diff
@@ -1,6 +1,8 @@
 enum BSApiPath {
   ROOT = '/',
   SPACES = '/spaces',
+  SPACE = '/spaces/',
```
Comment (what1s1ove) [inline shared/src/common/enums/api/bs-api-path.enum.ts]:
```suggestion
```
should be without slash at the end

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: shared/src/common/enums/api/bs-api-path.enum.ts (hunk 13 above)
Comment (Donymak) [reply shared/src/common/enums/api/bs-api-path.enum.ts]: but I created this enum to have slash in the end, without slash it won't work
image
So, how should I add a slash in route?

Comment (what1s1ove) [reply shared/src/common/enums/api/bs-api-path.enum.ts]: add a new api enum for spaces

File: shared/src/common/enums/api/bs-api-path.enum.ts [hunk 14]
```diff
@@ -1,6 +1,7 @@
 enum BSApiPath {
   ROOT = '/',
   SPACES = '/spaces',
+  SPACES_$ID = '/spaces/:id',
```
Comment (what1s1ove) [inline shared/src/common/enums/api/bs-api-path.enum.ts]:
```suggestion
```

File: shared/src/common/enums/api/spaces-api-path.enum.ts [hunk 15]
```diff
@@ -0,0 +1,5 @@
+enum SpacesApiPath {
+  ROOT = '/',
```
Comment (what1s1ove) [inline shared/src/common/enums/api/spaces-api-path.enum.ts:2]:
```suggestion
  ROOT = '/',
  $ID = '/:id'
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/api/api.ts [hunk 16]
```diff
@@ -72,6 +73,12 @@ const initApi: FastifyPluginAsync = async (fastify) => {
     },
     prefix: ApiPath.SLC,
   });
+  fastify.register(initSpacesApi, {
+    services: {
+      space,
+    },
+    prefix: ApiPath.SPACES,
+  });
```
Comment (what1s1ove) [inline backend/src/api/api.ts]:
```suggestion
```
it should be part of the BS api

File: backend/src/api/spaces/spaces.api.ts [hunk 17]
```diff
@@ -0,0 +1,35 @@
+import { FastifyPluginAsync, FastifyRequest } from 'fastify';
```
Comment (what1s1ove) [inline backend/src/api/spaces/spaces.api.ts]: remove this file. it should be part of the `BS` api

File: frontend/src/services/bs-api/bs-api.service.ts [hunk 18]
```diff
@@ -39,6 +40,15 @@ class BSApi {
     );
   }
 
+  public deleteSpace(id: string): Promise<boolean> {
+    return this.#http.load(
+      joinItems(this.#apiPrefix, BSApiPath.SPACES, SpacesApiPath.ROOT, id),
```
Comment (what1s1ove) [inline frontend/src/services/bs-api/bs-api.service.ts]:
```suggestion
      joinItems(this.#apiPrefix, BSApiPath.BS, BSApiPath.SPACES, SpacesApiPath.ROOT, id),
```

File: shared/src/common/enums/api/api-path.enum.ts [hunk 19]
```diff
@@ -5,6 +5,7 @@ enum ApiPath {
   EAM = '/eam',
   SC = '/sc',
   BS = '/bs',
+  SPACES = '/spaces',
```
Comment (what1s1ove) [inline shared/src/common/enums/api/api-path.enum.ts]:
```suggestion
```
should be part of `BS` api

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/api/bs/bs.api.ts [hunk 20]
```diff
@@ -63,6 +69,25 @@ const initBsApi: FastifyPluginAsync<Options> = async (fastify, opts) => {
       return rep.send(spaces).status(HttpCode.OK);
     },
   });
+
+  fastify.route({
+    method: HttpMethod.DELETE,
+    url: BSApiPath.SPACES + SpacesApiPath.$ID,
```
Comment (what1s1ove) [inline backend/src/api/bs/bs.api.ts]: use template strings

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by what1s1ove — APPROVED** [review]

**Review by Yaroslaww-1 — APPROVED** [review]

**Pull request merged**